### PR TITLE
Use MaybeUninit in VecDeque to remove the undefined behavior of slice

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

Ref https://github.com/rust-lang/rust/issues/74189. Adjust the code to follow the [doc.rust-lang.org/reference/behavior-considered-undefined.html](https://doc.rust-lang.org/reference/behavior-considered-undefined.html).

* Change the return type of `buffer_as_slice` from `&[T]` to `&[MaybeUninit<T>]`.
* Add some corresponding safety comments.

Benchmark results:

master 8d6f527530f4ba974d922269267fe89050188789

```rust
test collections::vec_deque::tests::bench_pop_back_100       ... bench:          47 ns/iter (+/- 1)
test collections::vec_deque::tests::bench_pop_front_100      ... bench:          50 ns/iter (+/- 4)
test collections::vec_deque::tests::bench_push_back_100      ... bench:          69 ns/iter (+/- 10)
test collections::vec_deque::tests::bench_push_front_100     ... bench:          72 ns/iter (+/- 6)
test collections::vec_deque::tests::bench_retain_half_10000  ... bench:     145,891 ns/iter (+/- 7,975)
test collections::vec_deque::tests::bench_retain_odd_10000   ... bench:     141,647 ns/iter (+/- 3,711)
test collections::vec_deque::tests::bench_retain_whole_10000 ... bench:     120,132 ns/iter (+/- 4,078)
```

This PR

```rust
test collections::vec_deque::tests::bench_pop_back_100       ... bench:          48 ns/iter (+/- 2)
test collections::vec_deque::tests::bench_pop_front_100      ... bench:          51 ns/iter (+/- 3)
test collections::vec_deque::tests::bench_push_back_100      ... bench:          73 ns/iter (+/- 2)
test collections::vec_deque::tests::bench_push_front_100     ... bench:          73 ns/iter (+/- 2)
test collections::vec_deque::tests::bench_retain_half_10000  ... bench:     131,796 ns/iter (+/- 5,440)
test collections::vec_deque::tests::bench_retain_odd_10000   ... bench:     137,563 ns/iter (+/- 3,349)
test collections::vec_deque::tests::bench_retain_whole_10000 ... bench:     128,815 ns/iter (+/- 3,289)
```

### CODE REVIEW

**Comment by RalfJung** [comment]
Sorry, I have a too large backlog of Rust things to do to review this, and have no familiarity with this code.

r? rust-lang/libs

**Comment by JmPotato** [comment]
A friendly ping~ @m-ou-se

**Comment by JmPotato** [comment]
Since I haven't received any feedback for a while, I want to try to re-invite a member to help me review, thanks for understanding.

r? rust-lang/libs

**Comment by JmPotato** [comment]
Since I haven't received any feedback for a while, I want to try to re-invite a member to help me review, thanks for understanding.

r? @dtolnay

File: library/alloc/src/collections/vec_deque/iter.rs [hunk 1]
```diff
@@ -70,18 +79,21 @@ impl<'a, T> Iterator for Iter<'a, T> {
...
-            // two slices: self.ring[self.tail..], self.ring[..self.head]
+            // Safety: two slices: self.ring[self.tail..], self.ring[..self.head] both are initialized.
             let (front, back) = self.ring.split_at(self.tail);
-            let mut back_iter = back.iter();
-            let res = back_iter.try_fold(init, &mut f);
-            let len = self.ring.len();
-            self.tail = (self.ring.len() - back_iter.len()) & (len - 1);
-            iter = front[..self.head].iter();
-            final_res = iter.try_fold(res?, &mut f);
+            unsafe {
```
Comment (m-ou-se) [inline library/alloc/src/collections/vec_deque/iter.rs:89]: Can you put the `unsafe` block aroudn only the expressions that need it? With a larger block like this it's a bit hard to see that the only unsafe code here are the `slice_assume_init_ref` calls.

File: library/alloc/src/collections/vec_deque/iter.rs [hunk 2]
```diff
@@ -142,17 +162,21 @@ impl<'a, T> DoubleEndedIterator for Iter<'a, T> {
...
         } else {
-            // two slices: self.ring[self.tail..], self.ring[..self.head]
+            // Safety: two slices: self.ring[self.tail..], self.ring[..self.head] both are initialized.
             let (front, back) = self.ring.split_at(self.tail);
-            let mut front_iter = front[..self.head].iter();
-            let res = front_iter.try_rfold(init, &mut f);
-            self.head = front_iter.len();
-            iter = back.iter();
-            final_res = iter.try_rfold(res?, &mut f);
+            unsafe {
```
Comment (m-ou-se) [inline library/alloc/src/collections/vec_deque/iter.rs:173]: (Same here.)

**Review by m-ou-se — COMMENTED** [review]
Nice, thank you for working on this! I left one small comment, but otherwise this looks good to me.

> Since I haven't received any feedback for a while, [..]

Unfortunately we're all quite busy and there's a lot of PRs to review, so it's not uncommon for a PR to sit still for a week or two. Unless it's a very critical issue, two weeks is probably a good time to send a ping or reassign it.

**Comment by JmPotato** [comment]
> Nice, thank you for working on this! I left one small comment, but otherwise this looks good to me.
> 
> > Since I haven't received any feedback for a while, [..]
> 
> Unfortunately we're all quite busy and there's a lot of PRs to review, so it's not uncommon for a PR to sit still for a week or two. Unless it's a very critical issue, two weeks is probably a good time to send a ping or reassign it.

Thanks for the explanation! It seems that I am too eager, xD. I have addressed the comment. PTAL.

**Comment by m-ou-se** [comment]
Thanks!

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r- 
failed in a rollup it seems:
https://github.com/rust-lang/rust/pull/94784#issuecomment-1063424379
@bors rollup=never (perf change)

**Comment by JmPotato** [comment]
> @bors r- failed in a rollup it seems: [#94784 (comment)](https://github.com/rust-lang/rust/pull/94784#issuecomment-1063424379) @bors rollup=never (perf change)

@matthiaskrgr @m-ou-se Hi, I updated the `Debug` implementation of `VecDeque` to keep compatible with the previous behavior. This should fix the falling test, PTAL again, thx!

**Comment by JmPotato** [comment]
A memory leak is found during my own test, it turns out that dropping a `MaybeUninit<T>` will never call `T`’s drop code. It is my responsibility to make sure `T` gets dropped if it got initialized. I will try to fix it later. ref [here](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#impl).

```shell
diff of stdout:

 [2, 2] Iter([2, 2], [])
-Iter([], [])
+Iter([2, 2], [])


The actual stdout differed from the expected stdout.
Actual stdout saved to /var/folders/x_/z7jrp4m90z948c1n3qnzfmt80000gn/T/compiletestHn7VSL/vecdeque.stage-id.stdout
normalized stderr:
The following memory was leaked: alloc4102 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}
alloc4151 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}
alloc8950 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}
alloc8999 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}

error: the evaluated program leaked memory

note: pass `-Zmiri-ignore-leaks` to disable this check

error: aborting due to previous error




The actual stderr differed from the expected stderr.
Actual stderr saved to /var/folders/x_/z7jrp4m90z948c1n3qnzfmt80000gn/T/compiletestHn7VSL/vecdeque.stage-id.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args vecdeque.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
command: "/Users/jmpotato/Code/rust/build/aarch64-apple-darwin/stage1-tools-bin/miri" "tests/run-pass/vecdeque.rs" "-L" "/var/folders/x_/z7jrp4m90z948c1n3qnzfmt80000gn/T/compiletestHn7VSL" "--target=aarch64-apple-darwin" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/var/folders/x_/z7jrp4m90z948c1n3qnzfmt80000gn/T/compiletestHn7VSL/vecdeque.stage-id" "-A" "unused" "--edition" "2018" "-Astable-features" "--sysroot" "/Users/jmpotato/Library/Caches/org.rust-lang.miri/HOST" "-Zmiri-tag-raw-pointers" "-L" "/var/folders/x_/z7jrp4m90z948c1n3qnzfmt80000gn/T/compiletestHn7VSL/vecdeque.stage-id.aux"
stdout:
------------------------------------------
[2, 2] Iter([2, 2], [])
Iter([2, 2], [])

------------------------------------------
stderr:
------------------------------------------
The following memory was leaked: alloc4102 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}
alloc4151 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}
alloc8950 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}
alloc8999 (Rust heap, size: 4, align: 4) {
    02 00 00 00                                     │ ....
}

{"message":"the evaluated program leaked memory","code":null,"level":"error","spans":[],"children":[],"rendered":"error: the evaluated program leaked memory\n\n"}
{"message":"pass `-Zmiri-ignore-leaks` to disable this check","code":null,"level":"note","spans":[],"children":[],"rendered":"note: pass `-Zmiri-ignore-leaks` to disable this check\n\n"}
{"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}

------------------------------------------

test [ui] run-pass/vecdeque.rs ... FAILED
```

**Comment by rylev** [comment]
The micro benchmarks indicate that this is largely a performance wash. Most benchmarks don't seem to show statistical difference and those that do are a mix of small regressions and improvements. It is surprising to see one very large improvement though. Given this is a correctness fix though and overall the perf changes leaned towards improvement, I think we can live with this.

@rustbot label: +perf-regression-triaged

**Pull request merged**